### PR TITLE
Fix Race in ClusterApplierService Shutdown

### LINKED ISSUE #62863: IndexRecoveryIT.testPeerRecoveryTrimsLocalTranslog is stuck sporadically
**Build scan**: 

on 7.x:

* https://gradle-enterprise.elastic.co/s/n6g3wfnbhpsw2
* https://gradle-enterprise.elastic.co/s/ihnyeqjamqaem
* https://gradle-enterprise.elastic.co/s/d4kf4qkmcr2ds

on master:

* https://gradle-enterprise.elastic.co/s/phxortxfty7tc
* https://gradle-enterprise.elastic.co/s/fwmfuivftffwy
* https://gradle-enterprise.elastic.co/s/2m5vt6geehu4u

**Repro line**:

```
./gradlew ':server:internalClusterTest' --tests "org.elasticsearch.indices.recovery.IndexRecoveryIT" -Dtests.seed=7C2C36E235B18ACB -Dtests.security.manager=true -Dtests.locale=en-US -Dtests.timezone=UTC -Druntime.java=8
```

**Reproduces locally?**: no

**Applicable branches**:

* master
* 7.x

**Failure history**:

According to [build stats](https://build-stats.elastic.co/app/kibana#/discover?_g=(refreshInterval:(pause:!t,value:0),time:(from:now-3M,mode:relative,to:now))&_a=(columns:!(class,test,branch),index:e58bf320-7efd-11e8-bf69-63c8ef516157,interval:auto,query:(language:lucene,query:'class:%22org.elasticsearch.indices.recovery.IndexRecoveryIT%22%20AND%20test:%22testPeerRecoveryTrimsLocalTranslog%22%20AND%20NOT%20branch:%22pull-request%22'),sort:!(time,desc))) this test has failed six times (excluding pull request builds); all of these errors have occured within the last two weeks.

**Failure excerpt**:

The failure manifests as suite timeout:

```
11:33:39   2> java.lang.Exception: Suite timeout exceeded (>= 1200000 msec).
11:33:39         at __randomizedtesting.SeedInfo.seed([7C2C36E235B18ACB]:0)
```

This might be a red herring but in the 7.x build failures we also see in the output of the active threads that the test is waiting for indexing to finish (this one is from the [most recent failure](https://gradle-enterprise.elastic.co/s/n6g3wfnbhpsw2)):

```
11:33:35   2>   23) Thread[id=3068, name=Thread-6, state=WAITING, group=TGRP-IndexRecoveryIT]
11:33:35   2>         at sun.misc.Unsafe.park(Native Method)
11:33:35   2>         at java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
11:33:35   2>         at java.util.concurrent.locks.AbstractQueuedSynchronizer.parkAndCheckInterrupt(AbstractQueuedSynchronizer.java:836)
11:33:35   2>         at java.util.concurrent.locks.AbstractQueuedSynchronizer.doAcquireSharedInterruptibly(AbstractQueuedSynchronizer.java:997)
11:33:35   2>         at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquireSharedInterruptibly(AbstractQueuedSynchronizer.java:1304)
11:33:35   2>         at org.elasticsearch.common.util.concurrent.BaseFuture$Sync.get(BaseFuture.java:259)
11:33:35   2>         at org.elasticsearch.common.util.concurrent.BaseFuture.get(BaseFuture.java:87)
11:33:35   2>         at org.elasticsearch.common.util.concurrent.FutureUtils.get(FutureUtils.java:56)
11:33:35   2>         at org.elasticsearch.action.support.AdapterActionFuture.actionGet(AdapterActionFuture.java:37)
11:33:35   2>         at org.elasticsearch.action.ActionRequestBuilder.get(ActionRequestBuilder.java:52)
11:33:35   2>         at org.elasticsearch.indices.recovery.IndexRecoveryIT.lambda$testPeerRecoveryTrimsLocalTranslog$42(IndexRecoveryIT.java:1802)
11:33:35   2>         at org.elasticsearch.indices.recovery.IndexRecoveryIT$$Lambda$4561/181390541.run(Unknown Source)
11:33:35   2>         at java.lang.Thread.run(Thread.java:748)
```

See also:

https://github.com/elastic/elasticsearch/blob/752590041e46fff97496b7487a1110581ac0593f/server/src/internalClusterTest/java/org/elasticsearch/indices/recovery/IndexRecoveryIT.java#L1800-L1802

### PR DESCRIPTION
**Pull request by original-brownbear**
The iteration over `timeoutClusterStateListeners` starts when the CS applier
thread is still running. This can lead to entries being added to it that never
get their listener resolved on shutdown and thus leak that listener as observed
in a stuck test in #62863 (in this particular test it's a failure to fail the shard on a stopping node that triggers waiting for the next cluster state for which the listener here can slip through the cracks).
Since `listener.onClose()` is idempotent we can just call it if we run into a stopped service
on the CS thread to avoid the race with certainty (because the iteration in `doStop` starts after
the stopped state has been set).

Closes #62863

### CODE REVIEW

**Comment by original-brownbear** [comment]
Jenkins run elasticsearch-ci/2 (unrelated x-pack)

File: server/src/main/java/org/elasticsearch/cluster/service/ClusterApplierService.java [hunk 1]
```diff
@@ -274,6 +274,10 @@ public void run() {
                     if (timeout != null) {
                         notifyTimeout.cancellable = threadPool.schedule(notifyTimeout, timeout, ThreadPool.Names.GENERIC);
                     }
+                    if (lifecycle.stoppedOrClosed()) {
```
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/cluster/service/ClusterApplierService.java:277]: Could we not do this 3 lines earlier? No need to schedule a task in this case.

**Review by henningandersen — APPROVED** [review]
LGTM

File: server/src/main/java/org/elasticsearch/cluster/service/ClusterApplierService.java (hunk 1 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/cluster/service/ClusterApplierService.java:277]: Right we could -> will do :)

**Comment by original-brownbear** [comment]
Jenkins run elasticsearch-ci/packaging-sample-windows

**Comment by original-brownbear** [comment]
Thanks Henning!

**Pull request merged**